### DISCUSSION TITLE
bcc, bpftrace: what are the ways to get contents from read syscall and get contents from a function in a program?

### DISCUSSION THREAD

**Discussion by wzis** [bpftrace/bpftrace#1731] [General]
I'm new to bcc and bpftrace, want to know the ways an attacker could get contents from read syscall and ways to get contents from a function like
int getpass(char *buf, int size)
in a program using bcc or bpftrace.

Tring to find a way to avoid such attacks.

Thanks

**Comment by fbs** [comment]
It is unclear to me what you want to achieve with this, can you elaborate?

For user space tracing we have uprobes that attach to a function. Additionally you can use watchpoints to do the same thing.
For kernel we have tracepoints, kprobes and kfuncs that can do it easily. 

But if you know at which address to look you can exfiltrate data from any probe as you can read memory adddresses.

tl;dr, if an attacker has access to the bpf syscall you're screwed.

**Reply by wzis** [reply]
Thanks for your reply: I want to see some example BCC or bpftrace scripts that can get the contents which the read syscall returns from any processes or a specific process ID, and when you know a program that has a function
int getpass(int fd, char *buf, int size)
and when that program runs, the scripts can get the contents in buf when the getpass function returned.
with those examples, I want to see if there are ways to detect and/or prevent it from happening.

**Reply by fbs** [reply]
> I want to see some example BCC or bpftrace scripts that can get the contents which the read syscall returns from any processes or a specific process ID

Take a look at the opensnoop.bt and bashreadline.bt or gethostlatency.bt tools, they're a good starting point. 

As for the pid filter, add `/pid == 1234/` as a predicate to filter by pid, where 1234 is the pid you want to trace.

>  I want to see if there are ways to detect and/or prevent it from happening.

The only way to prevent it is to not give unprivileged users access to the bpf syscall. For detection you could write a bpf program that traces the bpf syscall or the kernel functions that do the program loading/attaching.

**Reply by wzis** [reply]
I copied the bashreadline.bt, and changed it to

```
#!/usr/bin/env bpftrace
/*
 * bashreadline    Print entered bash commands from all running shells.
 *                 For Linux, uses bpftrace and eBPF.
 *
 * This works by tracing the readline() function using a uretprobe (uprobes).
 *
 * USAGE: bashreadline.bt
 *
 * This is a bpftrace version of the bcc tool of the same name.
 *
 * Copyright 2018 Netflix, Inc.
 * Licensed under the Apache License, Version 2.0 (the "License")
 *
 * 06-Sep-2018  Brendan Gregg   Created this.
 */

BEGIN
{
        printf("Tracing wzgetpass commands... Hit Ctrl-C to end.\n");
        printf("%-9s %-6s %s\n", "TIME", "PID", "COMMAND");
}

uretprobe:/usr/local/bin/wzappkey:wzgetpass
{
        time("%H:%M:%S  ");
        printf("%-6d %s\n", pid, str(retval));
}
```
When I run it, got an error:

```
root@ub2204:/# ./wzappkey.bt
Attaching 2 probes...
Could not resolve symbol: /usr/local/bin/wzappkey:wzgetpass
root@ub2204:/#
```

I have 

```
root@ub2204:/# ls -l /usr/local/bin/wzappkey
-rws--x--x 1 root daemon 124736 Mar 10 10:22 /usr/local/bin/wzappkey
root@ub2204:/#
```

So I believe by striping the symbol table from the program, it can effectively block this type of attack. 

And for the opensnoop.bt, I can see it can only get the argument passed in to the open, and the return value of the read system call to check whether it's successful or not, not the content read from the filename. So it's not what I wanted for verifying security software.

And the gethostlatency.bt doesn't seem to be suitable for my urpose.
Brendan wrote a sshsnoop using dtrace before, what I wanted is something similar but wrote in BCC or bpftrace, to see if BCC and bpftrace can steal the password the user entered for the software.

**Reply by fbs** [reply]
> So I believe by striping the symbol table from the program, it can effectively block this type of attack.

No, the kernel attaches on an address not a symbol. The symbol only makes life easier.

> Brendan wrote a sshsnoop using dtrace before, what I wanted is something similar but wrote in BCC or bpftrace, to see if BCC and bpftrace can steal the password the user entered for the software.

Read takes a `buf` argument which you can extract data from.

**Reply by wzis** [reply]
I need an example to learn this type of attack and try to think of a way to block it.

**Reply by fbs** [reply]
Again, it cannot be blocked. Once an attacker can load / attach bpf programs they can access everything. Thats why the bpf syscall is root user only and not namespaced, it just isn't safe.

As an example you can look at the refguide for the uprobe attach syntax to figure it out. If you want to know more about bpf internals you'll have to dig in the kernel.

**Reply by brendangregg** [reply]
Please, before anyone thinks the wrong thing: BPF is **not** making this attack possible. This is not a BPF attack or vulnerability. You can sniff passwords and plaintext using strace(1) or a debugger or ptrace or /proc/mem etc.

If you start saying that you've discovered a BPF security issue, then people become **misled** and start disabling BPF in their environments: which is the worst possible outcome for the BPF community. You're basically dropping a nuclear bomb on everything we've built.

BPF does make this style of attack easier to do. But it does not make it possible: it's already possible using debuggers.

**Comment by wzis** [comment]
Our software currently can detect original Orcale dtrace, strace, systemtap, /proc/pid/mem, /dev/mem, debugger, to prevent them to steal the password. That's why we want to know BPF could be used to steal password from processes running on the system provided the software when get password will immediately use it and remove it. There is one way to make BPF usable and not affect the security: give developer a way to disable/suspend it in real time when need to read password.

________________________________

**Reply by wzis** [reply]
In our mind, info can only be more secure when you have zero-trust computing software to help users to do the tasks, no one should be trusted, especially those with root privileges.

**Reply by danobi** [reply]
I'm curious how you are implementing this detection -- do you have any code links?

I'm also curious why the focus is on reading passwords. Most production environments I've worked in rely on certificates and one-time keys.

**Reply by wzis** [reply]
For example, with our software, we have detected the Xorg X-Window server software on RHEL 5.x could be used as a memory snooper as it opens /dev/mem whole life, so if someone modifies it for the purpose of memory snooping, no one would notice it, except our software has already given a warning.
Certificates stored on the system could be copied by any person with root access; one-time key is generated from a machine, and that machine's sysadmin could change the program. So, without zero-trust computing software to help, info can hardly be adequately protected. OS feature developers need to think about the issue, and always create a solution that not only will make troubleshooting easier, but also can let a process disable/suspend it in the case when critical info needs to be protected.

**Reply by wzis** [reply]
currently, we are using the following files to try block ftrace, kprobe, tracing:
/proc/sys/kernel/ftrace_enabled
/sys/kernel/debug/kprobes/enabled
/sys/kernel/debug/tracing/tracing_on
but they do not have effect on bpf tracing.

**Reply by danobi** [reply]
What would stop someone with root from writing "on" into all those files?

Edit: Because if you're going to just give up when an attacker has root (as you rightly should), then bpf is not relevant in this discussion as you need root to run bpf programs.

**Reply by wzis** [reply]
If these ones can effectively stop ftrace/kprobe/tracing, making currently ongoing ones discarded, then when the attacker with root privilege re-enable them, only partial info could be captured, can hardly capture full info.
We are trying to not give up when an attacker got root access, at least to make the attack be found earlier, not hiding in the background as the Linux fanotify will allow the attacker to hide critical config file change.

**Reply by danobi** [reply]
If the attacker has root, here's some things they could do:
* mount the real tracefs somewhere else and place some dummy files in `/sys/kernel/debug/tracing`
* install a kernel module that disables your security service
* replace the installed kernel with their own rootkit'd kernel
* simply install kprobes (and perhaps uprobes) via their kernel module (b/c kprobes have a module API)

And that's just what I've come up in a couple minutes. A motivated attacker could certainly do better.

I don't expect I'll convince you here so I'll just leave with this: disabling BPF does not make you any safer from attacks that steal passwords -- it only makes you **feel** safer.

**Reply by wzis** [reply]
We are using the fs type to find the location of those files, not the fixed path.

**Reply by wzis** [reply]
We provide multi-level protection for the systems: first users need to use our privilege delegation software CaclMgr, this one is much more secure than any other similar function software, will record command history reliably, prevent sysadmin to use another person's ID to run the PD software; then our software provides a set of tools to allow users to run other commands to have protection against original Oracle dtrace, strace, systemtap, debugger, ftrace, kprobe, memory snooper, tty keylogger and Trojan Horse attacks.

**Reply by wzis** [reply]
We know Linux has many security issues in design:
1. the /proc/kallsyms gives the attacker all kernel variable's addresses.
2. fanotify allows the attacker to hide critical file changes without being able to be detected.
3. the open source nature means anyone can download the source package to modify it and rebuild it to replace the installed one.
4. allow the kernel to be updated without the need for a reboot.
5. because the /proc/kallsyms, rootkit once loaded, can hide in the kernel.

**Reply by wzis** [reply]
BTW, we are not saying with our software, the system will be 100% secure, we are saying we want to make the attack hard to succeed and/or hard to hide. Using the normal software, there are too many problems, for example, using sudo for privilege delegation, by stealing the victim's OS password, then can use victim's ID to run sudo, which also mean, sysadmin can just temp replace the victim's /etc/shadow password, to use victim's ID to run sudo.

**Reply by wzis** [reply]
We have found a way to block the bpftrace to steal password from our program protected process. Hope the implementation way of BPF won't dramatically changed.

**Reply by wzis** [reply]
root@ub2204:/# date;./biostacks.bt ;date
Fri 19 Mar 2021 20:17:53 AEDT
Attaching 5 probes...
cannot attach kprobe, probe entry may not exist
Warning: could not attach probe kprobe:blk_start_request, skipping.
Tracing block I/O with init stacks. Hit Ctrl-C to end.
Killed
Fri 19 Mar 2021 20:17:53 AEDT

**Reply by fbs** [reply]
Let's keep the discussion relevant to bpftrace (this is bpftrace help/questions after all). Posting 'random' output from your security tool without context/details won't be beneficial to others. So please post what you did to achieve this so people can learn from it.

Also keep in mind that discussions on extra security mechanisms for/related BPF should be held in the BPF mailing lists. It's not in the scope of the bpftrace project. We just provide a tool that builds on the tools that Linux provides us, which can be used for good and for bad, just like strace&perf can.
